airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/pull_request/pipeline.py

```diff
@@ -0,0 +1,396 @@
+    branch_id: str
+    write: bool
+    input_title: str | None
+    input_body: str | None
+    changelog: bool
+    bump: str | None
+
+    title = "Create a pull request of changed files."  # type: ignore
+
+    def __init__(
+        self,
+        context: ConnectorContext,
+        message: str,
+        branch_id: str | None,
+        input_title: str | None,
+        input_body: str | None,
+        dry_run: bool,
+    ) -> None:
+        super().__init__(context)
+        self.message = message
+        self.branch_id = branch_id or default_branch_details(message)  # makes branch like: {branch_id}/{connector_name}
+        self.input_title = input_title
+        self.input_body = input_body
+        self.write = not dry_run
+
+    async def _run(self) -> StepResult:  # type: ignore
+
+        connector_files = await get_connector_changes(self.context)
+        if len(connector_files) == 0:
+            return StepResult(step=self, status=StepStatus.SKIPPED, stderr="No files modified in this connector.")
+
+        pull_request_number = await create_github_pull_request(
+            write=self.write,
+            context=self.context,
+            file_paths=connector_files,
+            branch_id=self.branch_id,
+            message=self.message,
+            input_title=self.input_title,
+            input_body=self.input_body,
+        )
+
+        return StepResult(step=self, status=StepStatus.SUCCESS, output={PULL_REQUEST_OUTPUT_ID: pull_request_number})
+
+
+async def get_connector_changes(context: ConnectorContext) -> Set[Path]:
+    logger = main_logger
+    all_modified_files = set(
+        transform_strs_to_paths(
+            await get_modified_files(
+                context.click_context.obj["git_branch"],
+                context.click_context.obj["git_revision"],
+                context.click_context.obj["diffed_branch"],
+                context.click_context.obj["is_local"],
+                context.click_context.obj["ci_context"],
+                context.click_context.obj["git_repo_url"],
+            )
+        )
+    )
+
+    directory = context.connector.code_directory
+    logger.info(f"    Filtering to changes in {directory}")
+    # get a list of files that are a child of this path
+    connector_files = set([file for file in all_modified_files if directory in file.parents])
+    # get doc too
+    doc_path = context.connector.documentation_file_path
+
+    if doc_path in all_modified_files:
+        connector_files.add(doc_path)
+
+    return connector_files
+
+
+def default_branch_details(message: str) -> str:
+    transformed = re.sub(r"\W", "-", message.lower())
+    truncated = transformed[:20]
+    data_bytes = message.encode()
+    hash_object = hashlib.sha256(data_bytes)
+    desc = f"{truncated}-{hash_object.hexdigest()[:6]}"
+    return desc
+
+
+@dataclass
+class ChangedFile:
+    path: str
+    sha: str | None
+
+
+# outputs a pull request number
+async def create_github_pull_request(
```

Comment:
I did not know that you could create refs and commits from the github client.

Would a simpler approach (which I used in the past on some adhoc scripts) be to:
For each connector, with `git` in `python`:
* Create a local branch
* Commit all files modified in the connector directory (+doc) on this branch
* (force) push to the remote repo
* Then open a PR with the github client

Would this address your comment below?

I kinda like to create local branches as developers could iterate on them post creation.

Another approach with containers would be to:
* Create a container with `git` and the github CLI
* Mount it the airbyte remote master branch state with dagger (https://dagger-io.readthedocs.io/en/sdk-python-v0.9.6/client.html#dagger.Client.git)
* Create a new branch
* Mount the connector directory
* Run git commands to stage, commit and push connectors files
* Run the `gh` command to open a PR

Reply:
I preferred this approach because it didn't change the local system.

Reply:
Let's keep it this way if it works. If we have issues with deleted files we could use my "with containers" approach later.